ArmPkg/Drivers/SmmVariablePei/SmmVariablePei.c

```diff
@@ -0,0 +1,225 @@
+  @param  VariableGuid          A pointer to an EFI_GUID that is the variable's GUID. The combination of
+                                VariableGuid and VariableName must be unique.
+  @param  Attributes            If non-NULL, on return, points to the variable's attributes.
+  @param  DataSize              On entry, points to the size in bytes of the Data buffer.
+                                On return, points to the size of the data returned in Data.
+  @param  Data                  Points to the buffer which will hold the returned variable value.
+                                May be NULL with a zero DataSize in order to determine the size of the buffer needed.
+
+  @retval EFI_SUCCESS           The variable was read successfully.
+  @retval EFI_NOT_FOUND         The variable was be found.
+  @retval EFI_BUFFER_TOO_SMALL  The DataSize is too small for the resulting data.
+                                DataSize is updated with the size required for
+                                the specified variable.
+  @retval EFI_INVALID_PARAMETER VariableName, VariableGuid, DataSize or Data is NULL.
+  @retval EFI_DEVICE_ERROR      The variable could not be retrieved because of a device error.
+
+**/
+EFI_STATUS
+EFIAPI
+PeiSmmGetVariable (
+  IN CONST  EFI_PEI_READ_ONLY_VARIABLE2_PPI  *This,
+  IN CONST  CHAR16                           *VariableName,
+  IN CONST  EFI_GUID                         *VariableGuid,
+  OUT       UINT32                           *Attributes,
+  IN OUT    UINTN                            *DataSize,
+  OUT       VOID                             *Data OPTIONAL
+  )
+{
+  EFI_STATUS                                Status;
+  UINTN                                     MessageSize;
+  EFI_MM_COMMUNICATE_HEADER                 *CommunicateHeader;
+  SMM_VARIABLE_COMMUNICATE_HEADER           *SmmVarCommsHeader;
+  SMM_VARIABLE_COMMUNICATE_ACCESS_VARIABLE  *SmmVarAccessHeader;
+  UINT8                                     *MmCommunicateBuffer;
+  UINT8                                     RequiredPages;
+  EFI_PEI_MM_COMMUNICATION_PPI              *MmCommunicationPpi;
+
+  // Check input parameters
+  if ((VariableName == NULL) || (VariableGuid == NULL) || (DataSize == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((*DataSize > 0) && (Data == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = PeiServicesLocatePpi (&gEfiPeiMmCommunicationPpiGuid, 0, NULL, (VOID **)&MmCommunicationPpi);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to locate PEI MM Communication PPI: %r\n", __FUNCTION__, Status));
+    return Status;
+  }
+
+  // Allocate required pages to send MM request
+  RequiredPages = ((sizeof (EFI_MM_COMMUNICATE_HEADER) + \
+                    sizeof (SMM_VARIABLE_COMMUNICATE_HEADER) + \
+                    sizeof (SMM_VARIABLE_COMMUNICATE_ACCESS_VARIABLE) + \
+                    StrSize (VariableName) + *DataSize) \
+                   /EFI_PAGE_SIZE) + 1;

```

Comment:
Any reason not to use `EFI_SIZE_TO_PAGES()` here?

Reply:
Agreed. This is updated.